### PR TITLE
[gcp][feat] Add metrics collection

### PR DESCRIPTION
**Pull request by 1101-1**
# Description

## Services where metrics collection has completed:

- [x] Instance
- [x] Volume
- [x] Database
- [x] Load Balancers
- [x] Serverless

# Code of Conduct

By submitting this pull request, I agree to follow the [code of conduct](https://inventory.fix.security/code-of-conduct).

### CODE REVIEW

File: plugins/gcp/fix_plugin_gcp/resources/base.py [hunk 1]
```diff
@@ -374,13 +409,13 @@ def post_process_instance(self, builder: GraphBuilder, source: Json) -> None:
         """
         pass
 
+    def collect_usage_metrics(self, builder: GraphBuilder) -> List:  # type: ignore
```
Comment (aquamatthias) [inline plugins/gcp/fix_plugin_gcp/resources/base.py:412]: Can we define the list type?

File: plugins/gcp/fix_plugin_gcp/resources/monitoring.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+                "name": "projects/{project}",
+                "aggregation_crossSeriesReducer": "REDUCE_NONE",
+                "aggregation_groupByFields": "[]",
+                "interval_endTime": utc_str(end_time),
+                "interval_startTime": utc_str(start_time),
+                "view": "FULL",
+                # set parametes below dynamically
+                # "aggregation_alignmentPeriod": None,
+                # "aggregation_perSeriesAligner": None,
+                # "filter": None,
```
Comment (aquamatthias) [inline plugins/gcp/fix_plugin_gcp/resources/monitoring.py:166]: commented out on purpose?

File: plugins/gcp/fix_plugin_gcp/config.py [hunk 3]
```diff
@@ -17,7 +16,7 @@ class GcpConfig:
         metadata={"description": "GCP services to exclude (default: none)"},
     )
     project_pool_size: int = field(
-        factory=num_default_threads,
+        default=64,
```
Comment (aquamatthias) [inline plugins/gcp/fix_plugin_gcp/config.py:19]: what is the reason for this change?

File: plugins/gcp/fix_plugin_gcp/resources/base.py [hunk 4]
```diff
@@ -275,12 +329,61 @@ def for_region(self, region: GcpRegion) -> GraphBuilder:
...
+            self.last_run_started_at,
             self.graph_nodes_access,
             self.graph_edges_access,
         )
 
 
+@frozen(kw_only=True)
+class MetricNormalization:
+    unit: MetricUnit
+    stat_map: frozendict[str, StatName] = frozendict(
```
Comment (aquamatthias) [inline plugins/gcp/fix_plugin_gcp/resources/base.py:343]: Is this a static map? Does it make sense to have it as property?

File: plugins/gcp/fix_plugin_gcp/config.py (hunk 3 above)
Comment (1101-1) [reply plugins/gcp/fix_plugin_gcp/config.py:19]: to make it same as in azure and aws

File: plugins/gcp/fix_plugin_gcp/collector.py [hunk 5]
```diff
@@ -131,6 +159,18 @@ def collect(self) -> None:
...
 
+    def collect_usage_metrics(self, builder: GraphBuilder) -> None:
+        for resource in builder.graph.nodes:
+            if isinstance(resource, GcpResource) and (mq := resource.collect_usage_metrics(builder)):
+                start_at = builder.created_at - builder.metrics_delta
+                region = cast(GcpRegion, resource.region())
+                try:
+                    rb = builder.for_region(region)
+                    result = monitoring.GcpMonitoringMetricData.query_for(rb, mq, start_at, builder.created_at)
+                    monitoring.update_resource_metrics(resource, result)
```
Comment (aquamatthias) [inline plugins/gcp/fix_plugin_gcp/collector.py:170]: No. This way, you fetch metrics one by one.
You should call this method, pass the resource and not wait for the result.
The called method can then spawn threads etc.

**Review by aquamatthias — CHANGES_REQUESTED** [review]

**Review by aquamatthias — DISMISSED** [review]
LGTM - Thanks!

File: plugins/gcp/fix_plugin_gcp/resources/base.py [hunk 6]
```diff
@@ -300,6 +402,16 @@ def _keys(self) -> Tuple[Any, ...]:
...
 
+    @property
+    def resource_raw_name(self) -> str:
+        """
+        Extracts the last segment of the GCP resource ID.
+
+        Returns:
+            str: The last segment of the resource ID (e.g., "function-1" from "projects/{project}/locations/{location}/functions/function-1").
+        """
+        return self.id.strip().split("/")[-1]
```
Comment (aquamatthias) [inline plugins/gcp/fix_plugin_gcp/resources/base.py:413]:
```suggestion
        return self.id..rsplit("/", maxsplit=1)[-1]
```

**Review by aquamatthias — DISMISSED** [review]

**Review by aquamatthias — APPROVED** [review]

**Pull request merged**